crates/edr_evm/src/blockchain/local.rs

```diff
@@ -215,6 +108,33 @@ where
     }
 }
 
+impl<ChainSpecT> LocalBlockchain<ChainSpecT>
+where
+    <ChainSpecT as RuntimeSpec>::LocalBlock: BlockReceipts<
+        Arc<ChainSpecT::BlockReceipt>,
+        Error = BlockchainErrorForChainSpec<ChainSpecT>,
+    >,
+    ChainSpecT:
+        GenesisBlockFactory<LocalBlock = <ChainSpecT as RuntimeSpec>::LocalBlock> + SyncRuntimeSpec,
+{
+    /// Constructs a new instance using the provided arguments to build a
+    /// genesis block.
+    #[cfg_attr(feature = "tracing", tracing::instrument(skip_all))]
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        genesis_diff: StateDiff,
+        chain_id: u64,
+        hardfork: <ChainSpecT as ChainHardfork>::Hardfork,
+        options: GenesisBlockOptions,
+    ) -> Result<Self, <ChainSpecT as GenesisBlockFactory>::CreationError> {
+        let genesis_block = ChainSpecT::genesis_block(genesis_diff.clone(), hardfork, options)?;
+
+        Ok(unsafe {
```

Comment:
`with_genesis_block_unchecked` requires that the block number be zero, but as far as I can tell nothing in this function guarantees that, so it should not be okay to call it. We should either assert that it is zero here, or propagate the requirement up by making something else unsafe (e.g. `GenesisBlockFactory` an `unsafe trait`). From previous conversations I assume this doesn't lead to memory unsafety, it's just about functional correctness, so I'd avoid propagating unsafe upwards and I'd also consider removing `unsafe` from `with_genesis_block_unchecked`.

Reply:
Or we could use the safe `with_genesis_block`.

Reply:
I'll use the safe version instead

Reply:
Addressed by https://github.com/NomicFoundation/edr/pull/1018/commits/15cb535131ea94d1f9863ebcfa67fd8da0b145c3